### PR TITLE
gh-112205: Support `@setter` annotation from AC

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-112205

### CODE REVIEW

File: Tools/clinic/clinic.py
Comment (corona10) [inline Tools/clinic/clinic.py:1982]: @erlend-aasland @AlexWaygood 
If you want to propose a better way for this, please let me know.

**Comment by corona10** [comment]
@colesbury @erlend-aasland @AlexWaygood 
The current implementation is not super pretty, so please leave feedback for this implementation.
I tried not to touch the overall architecture of AC as possible.

* Unify GET/SET method macro definition with macro under technique.
* Update AC to support `@setter` annotation.

File: Tools/clinic/clinic.py
Comment (corona10) [inline Tools/clinic/clinic.py:1998]: https://github.com/python/cpython/pull/112922/files#r1421696103
ditto

**Comment by corona10** [comment]
For the dev guide, I will send a PR once the PR is landed.

**Comment by corona10** [comment]
And also, we need to consider supporting PyDoc for getter/setter with the separated PR.
https://github.com/python/cpython/blob/384d6c1fe7e24d1c88f85c3f443e5420e54ca935/Modules/_io/textio.c#L181-L186

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        #if defined({getset_name}_GETSETDEF)
        #   undef {getset_name}_GETSETDEF
        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        #if defined({getset_name}_GETSETDEF)
        #   undef {getset_name}_GETSETDEF
        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #   define {getset_name}_GETSETDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
            fail("neither @getter nor @setter can define parameters")
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
            template_dict['getset_name'] = f.c_basename.upper()
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: The return value variable name has a template (`{return_value}`). It _should_ be possible to use it most places, but unfortunately, it is still hard-coded most places 😢 It should map correctly here, though:

```suggestion
            return_value_declaration = "int {return_value};"
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:1982]: I'm not sure it is a good idea to add the `value` param implicitly. OTOH, it does reduce clinic input boilerplate code; all setters have the same argument spec, so why not declare it implicitly?

Implicitly adding the param prevents being able to add a param docstring, or a custom param name. That _could_ be solved by introspection, but I really do not think the added complexity would be worth it.

Alternatively, you could add a check that requires the clinic input to explicitly declare the `value` parameter.

I think implicitly adding it is fine.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
                # Implicitly add the setter value parameter.
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:1998]: I think this is fine; we do similar stuff for `__init__` functions. Perhaps a comment could be helpful, though.

```suggestion
            # All setters return an int.
            template_dict['impl_return_type'] = 'int'
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py:5378]: How about one of these? (Ditto for the `fail(...)` below.)

```suggestion
            fail("only one of @getter or @setter can be used")
```

(I'm not super enthusiastic about my suggestion; perhaps @AlexWaygood can cook up something nicer 😃)

Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: How about this? (Ditto for the `fail(...)` below.)

```suggestion
            fail("only one of @getter or @setter can be used")
```

(I'm not super enthusiastic about my suggestion; perhaps @AlexWaygood can cook up something nicer 😃)

**Review by erlend-aasland — COMMENTED** [review]
Overall this looks good to me; thanks!

I would prefer to use a naming aligned with the names already used in the C API (`tp_getset`, `PyGetSetDef`); that is, consistently use `getset` and `GETSET` instead of `getsetter` and `GETSETTER`.

File: Tools/clinic/clinic.py
Comment (corona10) [reply Tools/clinic/clinic.py:1982]: Yeah, if the current implicit declaration is acceptable, I am fine.
But we may consider refactoring them at some moment, even if we declare it implicitly.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py:890]: Nitpick: I see the other preprocessor templates are using either 2 or 4 space indents. Let's align with the existing style before merging.

```suggestion
        #  undef {getset_name}_GETSETTDEF
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
        #endif
    """)
    SETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet("""
        #if defined({getset_name}_GETSETDEF)
        #  undef {getset_name}_GETSETDEF
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #  define {getset_name}_GETSETTERDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Nitpick: I see the other preprocessor templates are using either 2 or 4 space indents. Let's align with the existing style before merging.

```suggestion
        #  undef {getset_name}_GETSETTDEF
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
        #endif
    """)
    SETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet("""
        #if defined({getset_name}_GETSETDEF)
        #  undef {getset_name}_GETSETDEF
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #  define {getset_name}_GETSETTERDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        if f.kind in {GETTER, SETTER} and parameters:
```

**Comment by corona10** [comment]
Thank you @erlend-aasland, I 've updated PR PTAL

**Comment by erlend-aasland** [comment]
Thanks. I'm ok with this PR, but let's wait for @AlexWaygood's review before we land it.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        #  undef {getset_name}_GETSETDEF
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        #  define {getset_name}_GETSETDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```

**Review by erlend-aasland — COMMENTED** [review]
Oops, I probably introduced a typo with my recent suggestions; sorry 'bout that!

**Review by erlend-aasland — APPROVED** [review]
LGTM. Let's wait for Alex to chime in.

(cc. @colesbury, if you'd like to take a look.)

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: I think we could give a slightly nicer error message here:

```suggestion
            fail(f"@{f.kind.name.lower()} method cannot define parameters")
```

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: (ignorable nit)

```suggestion
        if f.kind in {GETTER, SETTER}:
```

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: From the user's perspective, I think these are actually two distinct errors, so it makes sense to give two different error messages:

```suggestion
        if self.kind is GETTER:
            fail("Cannot apply @getter twice to the same function!")
        if self.kind is SETTER:
            fail("Cannot apply both @getter and @setter to the same function!")
```

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]:
```suggestion
        if self.kind is SETTER:
            fail("Cannot apply @setter twice to the same function!")
        if self.kind is GETTER:
            fail("Cannot apply both @getter and @setter to the same function!")
```

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]:
```suggestion
                fail(f"@{self.kind.name.lower()} methods cannot define a return type")
```

**Review by AlexWaygood — COMMENTED** [review]
Thanks! A few small suggestions, mostly to do with error messages:

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [reply Tools/clinic/clinic.py]: It also looks like this failure path might not have any test coverage; could you add a test for it? :)

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [reply Tools/clinic/clinic.py]: It also looks like this failure path might not have any test coverage; could you add a test for it? :)

File: Tools/clinic/clinic.py
Comment (corona10) [inline Tools/clinic/clinic.py:5388]: @AlexWaygood How about this?

Comment (AlexWaygood) [reply Tools/clinic/clinic.py:5388]: Very nice!

File: Lib/test/test_clinic.py
Comment (AlexWaygood) [inline Lib/test/test_clinic.py:2239]: I think it makes sense to have this as a separate test method:

```suggestion
    def test_getter_and_setter_disallowed_on_same_function(self):
        dup_annotations = [("@getter", "@setter"), ("@setter", "@getter")]
```

File: Lib/test/test_clinic.py
Comment (AlexWaygood) [inline Lib/test/test_clinic.py]: To fix the lint failure:

```suggestion
                expected_error = "Cannot apply both @getter and @setter to the same function!"
```

File: Lib/test/test_clinic.py
Comment (AlexWaygood) [inline Lib/test/test_clinic.py:2221]: It doesn't look like the `expected_error` variable here is actually used; did you mean to do this? :)

```suggestion
                expected_error = f"{annotation} method cannot define parameters"
                self.expect_failure(block, expected_error)
```

Comment (corona10) [reply Lib/test/test_clinic.py:2221]: Thank you for catching it!

**Review by AlexWaygood — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
